File: src/services/update.ts
import { invoke } from '@tauri-apps/api/core';

export interface UpdateUrls {
  // Windows 平台
  windows?: string; // 通用 Windows 安装包
  windows_exe?: string; // Windows .exe 安装包
  windows_msi?: string; // Windows .msi 安装包

  // macOS 平台
  macos?: string; // 通用 macOS 安装包
  macos_dmg?: string; // macOS .dmg 安装包

  // Linux 平台
  linux?: string; // 通用 Linux 安装包
  linux_deb?: string; // Debian/Ubuntu .deb 包
  linux_rpm?: string; // RedHat/CentOS .rpm 包
  linux_appimage?: string; // Linux AppImage 包

  // 通用版本
  universal?: string; // 跨平台通用版本
}

export interface UpdateInfo {
  current_version: string;
  latest_version: string;
  has_update: boolean;
  update_url?: string;
  update?: UpdateUrls;
  release_notes?: string;
  file_size?: number;
  required: boolean;
}

export interface DownloadProgress {
  downloaded_bytes: number;
  total_bytes: number;
  percentage: number;
  speed?: number;
  eta?: number;
}

export interface PlatformInfo {
  os: string;
  arch: string;
  is_windows: boolean;
  is_macos: boolean;
  is_linux: boolean;
}

export interface PackageFormatInfo {
  platform: string;
  preferred_formats: string[];
  fallback_format: string;
}

export enum UpdateStatus {
  Idle = 'Idle',
  Checking = 'Checking',
  Available = 'Available',
  Downloading = 'Downloading',
  Downloaded = 'Downloaded',
  Installing = 'Installing',
  Installed = 'Installed',
  Failed = 'Failed',
  Rollback = 'Rollback',
  RolledBack = 'RolledBack',
}

// 检查应用更新
export async function checkForAppUpdates(): Promise<UpdateInfo> {
  return await invoke<UpdateInfo>('check_for_app_updates');
}

// 下载应用更新
export async function downloadAppUpdate(url: string): Promise<string> {
  return await invoke<string>('download_app_update', { url });
}

// 安装应用更新
export async function installAppUpdate(updatePath: string): Promise<void> {
  return await invoke<void>('install_app_update', { updatePath });
}

// 获取更新状态
export async function getAppUpdateStatus(): Promise<UpdateStatus> {
  return await invoke<UpdateStatus>('get_app_update_status');
}

// 回滚更新
export async function rollbackAppUpdate(): Promise<void> {
  return await invoke<void>('rollback_app_update');
}

// 获取当前版本
export async function getCurrentAppVersion(): Promise<string> {
  return await invoke<string>('get_current_app_version');
}

// 重启应用进行更新
export async function restartAppForUpdate(): Promise<void> {
  return await invoke<void>('restart_app_for_update');
}

// 获取平台信息
export async function getPlatformInfo(): Promise<PlatformInfo> {
  return await invoke<PlatformInfo>('get_platform_info');
}

// 获取推荐的包格式信息
export async function getRecommendedPackageFormat(): Promise<PackageFormatInfo> {
  return await invoke<PackageFormatInfo>('get_recommended_package_format');
}

// 监听下载进度事件
export function onUpdateDownloadProgress(callback: (progress: DownloadProgress) => void) {
  // 检查是否在Tauri环境中
  if (typeof window !== 'undefined' && (window as any).__TAURI__) {
    return (window as any).__TAURI__.event.listen('update-download-progress', (event: any) => {
      if (event && event.payload) {
        callback(event.payload as DownloadProgress);
      }
    });
  }

  // 如果不在Tauri环境中，返回一个空函数（保持API一致）
  return Promise.resolve(() => undefined);
}

File: src-tauri/src/models/update.rs
use serde::{Deserialize, Serialize};
use std::path::PathBuf;

/// 更新信息
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateInfo {
    pub current_version: String,
    pub latest_version: String,
    pub has_update: bool,
    pub update_url: Option<String>,
    pub update: Option<UpdateUrls>, // 所有平台的更新URL
    pub release_notes: Option<String>,
    pub file_size: Option<u64>,
    pub required: bool, // 是否为强制更新
}

/// 更新状态
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
pub enum UpdateStatus {
    #[default]
    Idle, // 空闲状态
    Checking,       // 检查更新中
    Available,      // 有可用更新
    Downloading,    // 下载中
    Downloaded,     // 下载完成
    Installing,     // 安装中
    Installed,      // 安装完成，等待重启
    Failed(String), // 更新失败
    Rollback,       // 回滚中
    RolledBack,     // 回滚完成
}

/// 下载进度
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DownloadProgress {
    pub downloaded_bytes: u64,
    pub total_bytes: u64,
    pub percentage: f32,
    pub speed: Option<u64>, // bytes per second
    pub eta: Option<u32>,   // estimated time remaining in seconds
}

/// 下载任务信息
#[derive(Debug, Clone)]
pub struct DownloadTask {
    pub url: String,
    pub file_path: PathBuf,
    pub total_size: Option<u64>,
    pub downloaded: u64,
    pub start_time: std::time::Instant,
}

/// 平台信息
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PlatformInfo {
    pub os: String,
    pub arch: String,
    pub is_windows: bool,
    pub is_macos: bool,
    pub is_linux: bool,
}

/// 包格式信息
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PackageFormatInfo {
    pub platform: String,
    pub preferred_formats: Vec<String>,
    pub fallback_format: String,
}

/// 更新配置
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateConfig {
    pub auto_check: bool,
    pub check_interval_hours: u32,
    pub download_in_background: bool,
    pub auto_install: bool,
}

impl Default for UpdateConfig {
    fn default() -> Self {
        Self {
            auto_check: true,
            check_interval_hours: 24,
            download_in_background: true,
            auto_install: false,
        }
    }
}

/// 镜像站API响应
#[derive(Debug, Deserialize)]
pub struct UpdateApiResponse {
    pub version: String,
    pub update: UpdateUrls,
    pub release_notes: Option<String>,
    pub required: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateUrls {
    // Windows 平台
    pub windows: Option<String>,     // 通用 Windows 安装包
    pub windows_exe: Option<String>, // Windows .exe 安装包
    pub windows_msi: Option<String>, // Windows .si 安装包

    // macOS 平台
    pub macos: Option<String>,     // 通用 macOS 安装包
    pub macos_dmg: Option<String>, // macOS .dmg 安装包

    // Linux 平台
    pub linux: Option<String>,          // 通用 Linux 安装包
    pub linux_deb: Option<String>,      // Debian/Ubuntu .deb 包
    pub linux_rpm: Option<String>,      // RedHat/CentOS .rpm 包
    pub linux_appimage: Option<String>, // Linux AppImage 包

    // 通用包（如果有的话）
    pub universal: Option<String>, // 跨平台通用包
}
